Add alternative hankel definition; handle k=0 input; restructuring

MuellerSeb opened the pull request:
This PR addresses multiple points:
- adding the alternative hankel transform definition (https://en.wikipedia.org/wiki/Hankel_transform#Alternative_definition)
- handling the input of ``k=0`` by explicit integration
- better handling of the FourierTransform with using the alternative hankel definition
- restructuring:
  - ~~using _f again to call function correctly~~
  - replacing ``_x_power`` with ``_r_power`` ~~and using it in _f~~
  - bloated methods in separate ``tools.py`` for cleanup
- adding test for the alternative hankel definition

MuellerSeb commented:
I still need to fix a little bug and care about the coverage.

MuellerSeb commented:
Should be ready for review now.

steven-murray commented on hankel/hankel.py:
```diff
@@ -179,40 +125,69 @@ def transform(self, f, k=1, ret_err=True, ret_cumsum=False, inverse=False):
+        # care about k = 0
+        if np.any(k_0):
+            # limit of J(nu, 0) considering powers of k
+            lim_k_pow = 0.5 if self.alt else 0  # in alt. def sqrt(k) involved
```
what does `lim_k_pow` represent?

MuellerSeb replied:
I will rename this to ``alt_pow``. An explanation will follow.

steven-murray commented on hankel/hankel.py:
```diff
@@ -179,40 +125,69 @@ def transform(self, f, k=1, ret_err=True, ret_cumsum=False, inverse=False):
+        if np.any(k_0):
+            # limit of J(nu, 0) considering powers of k
+            lim_k_pow = 0.5 if self.alt else 0  # in alt. def sqrt(k) involved
+            lim_r_pow = self._r_power + lim_k_pow + self.nu
```
what does `lim_r_pow` represent?

MuellerSeb replied:
This is the resulting power of r in the integral for the limit case of k -> 0.

steven-murray commented on hankel/hankel.py:
```diff
@@ -179,40 +125,69 @@ def transform(self, f, k=1, ret_err=True, ret_cumsum=False, inverse=False):
+            # limit of J(nu, 0) considering powers of k
+            lim_k_pow = 0.5 if self.alt else 0  # in alt. def sqrt(k) involved
+            lim_r_pow = self._r_power + lim_k_pow + self.nu
+            nu_th = self._k_power - 1 - lim_k_pow  # threshold
```
what does `nu_th` represent?

MuellerSeb replied:
This is the threshold for nu determining the behavior of the integral for k -> 0.

MuellerSeb commented on hankel/tools.py:
```diff
@@ -0,0 +1,289 @@
+    -----
+    J(nu, x) is approximately (x/2)^nu / Gamma(nu+1) for small x.
+    """
+    if alt:
```
You are right, that in the "alt" case, this kernel (also the better name, I will adopt that) is multiplied with ``sqrt(x)``. Doing this in one line would introduce numerical problems from nu= +- 1/2, where the sqrt(x) cancels out. But these cases are interesting for the SFT, where we result in simple sin and cos functions, that can be savely evaluated at 0.

steven-murray commented on hankel/hankel.py:
```diff
@@ -179,40 +125,69 @@ def transform(self, f, k=1, ret_err=True, ret_cumsum=False, inverse=False):
+
+            if np.isclose(self.nu, nu_th):
+                ret[k_0] = quad(integrand, 0, np.inf)[0] * lim * norm
+            elif self.nu < nu_th:
```
As far as I can tell, if `nu!=0` then the answer is always zero?

MuellerSeb replied:
I will give an explanation on that as a comment.

MuellerSeb commented on hankel/hankel.py:
```diff
@@ -35,128 +45,66 @@ class HankelTransform(object):
     h : float, optional
         The step-size of the integration.
+    alt : bool, optional
+        State if the alternative definition of the hankel transform
```
Done.

MuellerSeb commented:
Here some details on the calculations:
[pr30details.pdf](https://github.com/steven-murray/hankel/files/3942577/pr30details.pdf)

Keypoint is the limiting case of J_nu:
![lim](https://latex.codecogs.com/gif.latex?J_%7B%5Cnu%7D%28k%29%5Capprox%20%5Cfrac%7B%28%5Cfrac%7Bk%7D%7B2%7D%29%5E%5Cnu%7D%7B%5CGamma%28%5Cnu&plus;1%29%7D)

See here: https://dlmf.nist.gov/10.7#ii

MuellerSeb commented:
One last thing to argue about could be the transition from using ``r_power`` instead of ``x_power``, since this introduces a bit more calculation. In ``_f`` the outer product of x and k is used, so len(k) times more multiplications are performed.

MuellerSeb commented:
And a second thought: The np.inf case should return np.nan instead I think. Since only one factor of the integrand is inf, the sign would depend on the given function f. If both signs occur in the function, this limit would be hard to determine numerically. also for f=0 we wouldn't result in np.inf.

steven-murray commented:
Thanks for creating the doc. I *think* we don't have to worry about the limit as k goes to zero, as we are not doing any integrals over k. When k=0, the substitution of `x=kr` doesn't make sense any more anyway, as the limit of `r=\infty` is undefined for `x`. So you have to go back to the original definition of the transform itself (in which the `k_power` is zero). You have to use the definition of the Hankel transform, and the definition of the SFT independently and substitute `k=0`. For the latter, no hankel transform is even required, as it just turns into the integral over `f(r)`, and it is well-defined (so long as that integral converges). Does that make sense or am I missing something?

MuellerSeb commented:
As written in the details, that is what I do. To calculate the limit for k -> 0, I use the **original** integral. Maybe I should use the variable b from the documentation directly as ``k_power``, so no confusion comes up. For k -> 0, you are right, an integral with only f(r) will come as a result, but the factor from the limit of J(nu, kr)*k^b for k->0 has to be considered and that is what I have done. The limit depends on nu and b. The nice thing about this approach is, that we **can** handle the hankel transform and the SFT at the same time, since we only have to adjust the powers of r and k in the integral.

I know, that the original ``k_power=1`` comes from the substitution x=kr, that is why a ``-1`` occurs in the calculation for ``nu_th``.

The limit is also shown to be correct with the test-cases.

## My proposal would be now:
- to make things clearer use ``a`` from the detail pdf as ``r_power`` and ``b`` as ``k_power``
- set the result for k=0 to nan if the kernel converges to inf
- the multiplication with r^a should be done by multiplying with x^a in the ``_get_series`` method

steven-murray commented:
I've written up a more comprehensive doc here: https://www.overleaf.com/read/cdtwvhvnsrrj. Let me know what you think!

MuellerSeb commented:
Thanks for that wrap up. I totally agree with your calculations. And they are in line with mine. In equation (3) you have to distinguish 2 cases for nu != 0, either nu > 0 or < 0 (since nu only has to be >= -1/2). For nu > 0 the result is 0 as you have written. But for nu < 0 the result is undefined because of the limiting case of J_nu:
![lim](https://latex.codecogs.com/gif.latex?J_%7B%5Cnu%7D%28k%29%5Capprox%20%5Cfrac%7B%28%5Cfrac%7Bk%7D%7B2%7D%29%5E%5Cnu%7D%7B%5CGamma%28%5Cnu&plus;1%29%7D)

I only introduced the possibility to add r^A / k^B to the integrand, therefore the cases are not simply nu=0, >0 and <0. That is why I introduced the threshold ``nu_th`` (which is 0 in the default case that you have written).

The integral in (6) is the same in my calculations, since ``lim_r_pow`` will result in ``n-1``, so the integrand is ``f(r)*r^(n-1)``, and the limit of the J(nu) will be the correct term (the gamma pi thing). So we don't have to treat hankel and SFT separately ;-)

steven-murray commented:
Yup, just went through it again and realised you're right. Sorry to prolong the process! I agree with your plan to finish it off.

MuellerSeb commented:
No problem! :-) It's always better to double-check!

MuellerSeb commented:
Should be ready now. Also reverting the multiplication with r is probably more consistent with the get_h routine, since the _get_series method is called there.

MuellerSeb commented:
@steven-murray thanks for the approval! I will add the additions to the changelog.

I got 2 points more, too:
1. these lines are a bit arbitrary:
   ```python
   # replace k=0 with 1
   knorm[k_0] = 1
   k[k_0] = 1
   ```
   The resulting ``err`` and ``cumsum`` for k=0 are then calculated for k=1. I don't know what you are thinking about that? We could use ``kn0 = k[np.invert(k_0)]`` in the calculations and populate ``err`` with 0 at k_0 and ``cumsum`` can be constantly the resulting value for k=0.
2. in ``_get_series`` the following line could be simplified:
   ```python
   return np.pi * self.w * fres * self.kernel * self.dpsi
   ```
   The part ``np.pi * self.w * self.kernel * self.dpsi`` could be calculated **once** in ``__init__``, but we would need the powers for r and k as input for ``__init__`` (since in SFT they are set after calling the ``__init__`` of the parent class). I would suggest another input parameter called ``powers``, that is ``None`` by default (resulting in the current powers), or it could be a dict with keywords "r" and "k", providing the powers.

steven-murray commented:
Some good ideas!

1. For `k=0`, I guess `err` should be zero if the output is zero, otherwise it should be the output of `quad` for the error. And the `cumsum` should probably be what you suggest. Along those lines, I think we should be calling `_get_series(f, k[~k0]), and then inserting the `k=0` back in if it exists (takes less computations).
2. I don't really like passing the powers into the `__init__`, it seems like they should be properties of the class, not open to user input so much. However, we could make the `np.pi * self.w * self.kernel * self.dpsi` the result of a `cached_property`?

MuellerSeb commented:
@steven-murray 
I implemented these suggestions. Calculations for k=0 and k>0 are now separated.
And I created a property called ``_series_fac`` that caches the factors for the series. It has to be called once, then it will return the stored value.
I also updated the Changlog. I added an unreleased section (can be replaced with the next version number)

That begs for a 1.0 release :-)

Cheers,
Sebastian

steven-murray commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,18 @@
+**Bugfixes**
+
+- saver calculation of xrange_approx
+- caching of series factors for faster calculations
```
this is an enhancement!

MuellerSeb commented:
@steven-murray : Done. I also bumped version to 1.0.0.dev0 (development status).

And I don't have write access to this repository, so you have to merge it. Thanks for your patience. And all modifications are also a bit selfish, since one of my packages depends on this ;-)

steven-murray commented:
Perfect, thanks!
